Added problem and solution for cf-1713E (#3086)
cpinitiative/usaco-guide: solutions/cf-1713E.mdx

---
id: cf-1713E
source: Codeforces Div 2
title: Cross Swapping
author: Aakash Gokhale
---

## Solution

[Official Analysis](https://codeforces.com/blog/entry/105419)

Notice that $a[i][j]$ can only be swapped with $a[j][i]$ no matter which operations are performed.

In addition, we want to lexicographically sort the array. Thus, we can iterate and greedily process the array.

> For all $j$ such that $i < j$ (we only need to consider each pair once), we can determine whether we want to swap them or keep them the same.

Reviewer: ```suggestion
For all $j$ such that $i < j$ (we only need to consider each pair once and keeping $i < j$ will ensure this), we can determine whether we want to swap them or keep them the same.
```

As we are iterating greedily through the array, we can merge ($i$ and $j$) and ($i + n$ and $j + n$) if we don't want to swap. This works because if we swap neither or swap both the arrangement remains the same.
If we do want to swap we can merge ($i$ and $j + n$) and ($i + n$ and $j$). This works because if we swap exactly one of these, then the arrangement will change. 

After this we can construct the answer.

We can iterate over the entire grid. If it is a diaganal it will never change. If $a[i][j] == a[j][i]$ then we can manually set them to normal (since we didn't merge them in the disjoint set union when comparing them the states might have not been merged). If it is not a diaganol we can check whether $i$ and $j$ are merged in the disjoint set union. If they aren't then we will swap $a[i][j]$ and $a[j][i]$. Otherwise we will keep them the same.

## Implementation

**Time Complexity:** $\mathcal{O}(N^2)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

using vi = vector<int>;

struct DSU {
	vi e;
	DSU(int n) {
		e = vi(n, -1);
	}

> 	int get(int x) {
> 		return e[x] < 0 ? x : e[x] = get(e[x]);
> 	}

Reviewer: ```suggestion
	int get(int x) {
		if (e[x] < 0) return x;
		else {
			e[x] = get(e[x]);
			return e[x];
		}
	}
```

	bool same(int a, int b) {
		return get(a) == get(b);
	}
	int size(int x) {
		return -e[get(x)];
	}
	bool unite(int x, int y) {
		x = get(x); y = get(y);
		if (x == y) return false;
		if (e[x] > e[y]) swap(x, y);
		e[x] += e[y]; e[y] = x;
		return true;
	}
};

void solve() {
	int n;
	cin >> n;
	int a[n][n];
	for (int i = 0; i < n; i++)
	  for (int j = 0; j < n; j++) cin >> a[i][j];
	DSU dsu(2 * n);
	for (int i = 0; i < n; i++) {
		for (int j = i + 1; j < n; j++) { // Only check each pair once and diagnals don't matter
			// If they are the same we should not merge either way because it could affect later merges
			if (a[i][j] == a[j][i]) continue;
			else if (a[i][j] > a[j][i]) {
				// Try to swap the elements

> 				if (dsu.get(i) != dsu.get(j)) { // Check if the states are already merged the opposite way

Reviewer: LINE LENGTH EXCEEDS 80 CHARACTERS

Author: fixed

					dsu.unite(i, j + n);
					dsu.unite(j, i + n);
				}
			} else	{
				// Try to keep the elements the same
				if (dsu.get(i) != dsu.get(j + n)) { // Check if the states are already merged the opposite way
					dsu.unite(i, j);
					dsu.unite(i + n, j + n);
				}
			}
		}
	}
	// Construct the answer
	int ans[n][n];
	for (int i = 0; i < n; i++) {
		for (int j = i; j < n; j++) {
			/* If a[i][j] is part of a diagonal or the pair is equal (a[i][j] == a[j][i]),
			 * manually set it to be the same as the original array
			 */
			if (i == j) ans[i][j] = a[i][j];
			else if (a[i][j] == a[j][i]) ans[i][j] = ans[j][i] = a[i][j];
			else { // Otherwise check whether the i and j states are merged
				if (dsu.get(i) == dsu.get(j)) {
					ans[i][j] = a[i][j];
					ans[j][i] = a[j][i];
				} else {
					ans[i][j] = a[j][i];
					ans[j][i] = a[i][j];
				}
			}
		}
	}

>   // Print the answer

Reviewer: ```suggestion
	// Print the answer
```

---
id: cf-1713E
source: Codeforces Div 2
title: Cross Swapping
author: Aakash Gokhale
---

## Solution

[Official Analysis](https://codeforces.com/blog/entry/105419)

Notice that $a[i][j]$ can only be swapped with $a[j][i]$ no matter which operations are performed.

In addition, we want to lexicographically sort the array. Thus, we can iterate and greedily process the array.

For all $j$ such that $i < j$ (we only need to consider each pair once and keeping $i < j$ will ensure this), we can determine whether we want to swap them or keep them the same.

If $a[i][j] > a[j][i]$, we want to swap them. If $a[i][j] < a[j][i]$ we want to keep them the same. If they are equal it doesn't matter.

However, it is possible that not all conditions can work together with each other and swapping something means we can't swap something else later on.

> We can use the [Amogus Trick](https://codeforces.com/blog/entry/106407) to represent this problem. Create a disjoint set union of $2n$ nodes where node $i$ represents the state where the operation is performed on $i$ and $i + n$ represents the state where operation is not performed on $i$.

Reviewer 2: is there no other way to solve this problem?
because if the only way to solve this is with some meme-a** trick, i don't think this is a good problem

Reviewer 2: or unless you can give a nice intuitive method for arriving at this trick

Reviewer 2: i wouldn't mention the trick then if it isn't really needed

Author: fixed

After this we can construct the answer.

We can iterate over the entire grid. If it is a diaganal it will never change. If $a[i][j] == a[j][i]$ then we can manually set them to normal (since we didn't merge them in the disjoint set union when comparing them the states might have not been merged). If it is not a diaganol we can check whether $i$ and $j$ are merged in the disjoint set union. If they aren't then we will swap $a[i][j]$ and $a[j][i]$. Otherwise we will keep them the same.

## Implementation

**Time Complexity:** $\mathcal{O}(N^2)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

using vi = vector<int>;

struct DSU {
	vi e;
	DSU(int n) {
		e = vi(n, -1);
	}
	
	int get(int x) {
		if (e[x] < 0) return x;
		else {
			e[x] = get(e[x]);
			return e[x];
		}
	}
	
	bool same(int a, int b) {
		return get(a) == get(b);
	}
	
	int size(int x) {
		return -e[get(x)];
	}
	
	bool unite(int x, int y) {
		x = get(x); y = get(y);
		if (x == y) return false;
		if (e[x] > e[y]) swap(x, y);
		e[x] += e[y]; e[y] = x;
		return true;
	}
};

void solve() {
	int n;
	cin >> n;
	int a[n][n];
	for (int i = 0; i < n; i++)

> 	  for (int j = 0; j < n; j++) cin >> a[i][j];

Reviewer 2: sus

Reviewer 2: braces + tabs

Author: added

	DSU dsu(2 * n);
	for (int i = 0; i < n; i++) {
		// Only check each pair once and diagnals don't matter
		for (int j = i + 1; j < n; j++) {
			// If they are the same we should not merge either way
			// because it could affect later merges
			if (a[i][j] == a[j][i]) continue;
			else if (a[i][j] > a[j][i]) {
				// Swap the elements if the states are not already merged the opposite way
				if (dsu.get(i) != dsu.get(j)) {
					dsu.unite(i, j + n);
					dsu.unite(j, i + n);
				}
			} else	{
				// Keep the elements the same if the states
				// are not already merged the opposite way
				if (dsu.get(i) != dsu.get(j + n)) {
					dsu.unite(i, j);
					dsu.unite(i + n, j + n);
				}
			}
		}
	}

> 	// Construct the answer

Reviewer 2: separate relevant code segments w/ whitespace

Author: fixed

		e[x] += e[y]; e[y] = x;
		return true;
	}
};

void solve() {
	int n;
	cin >> n;
	int a[n][n];
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			cin >> a[i][j];
		}
	}
	
	DSU dsu(2 * n);
	
	for (int i = 0; i < n; i++) {
		// Only check each pair once and diagnals don't matter
		for (int j = i + 1; j < n; j++) {
			// If they are the same we should not merge either way
			// because it could affect later merges
			if (a[i][j] == a[j][i]) continue;
			else if (a[i][j] > a[j][i]) {
				// Swap the elements if the states are not already merged the opposite way
				if (dsu.get(i) != dsu.get(j)) {
					dsu.unite(i, j + n);
					dsu.unite(j, i + n);
				}
			} else	{
				// Keep the elements the same if the states
				// are not already merged the opposite way
				if (dsu.get(i) != dsu.get(j + n)) {
					dsu.unite(i, j);
					dsu.unite(i + n, j + n);
				}
			}
		}
	}
	
	// Construct the answer
	int ans[n][n];
	for (int i = 0; i < n; i++) {
		for (int j = i; j < n; j++) {
			/* If a[i][j] is part of a diagonal or the pair is equal (a[i][j] == a[j][i]),
			 * manually set it to be the same as the original array
			 */
			if (i == j) ans[i][j] = a[i][j];

> 			else if (a[i][j] == a[j][i]) ans[i][j] = ans[j][i] = a[i][j];

Author: fixed

			else { // Otherwise check whether the i and j states are merged
				if (dsu.get(i) == dsu.get(j)) {
					ans[i][j] = a[i][j];
					ans[j][i] = a[j][i];
				} else {
					ans[i][j] = a[j][i];
					ans[j][i] = a[i][j];
				}
			}
		}
	}
	
	// Print the answer
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			cout << ans[i][j] << " ";
		}
		cout << endl;
	}
}

int main() { 

> 	ios::sync_with_stdio(false); cin.tie(nullptr);

Author: Yes

> 	ios::sync_with_stdio(false); cin.tie(nullptr);

Author: Even if it wasn't necessary, shouldn't you put fast io regardless?

> 	ios::sync_with_stdio(false); cin.tie(nullptr);

Reviewer 2: ? no?

> 	ios::sync_with_stdio(false); cin.tie(nullptr);

Reviewer 2: bump

> 	ios::sync_with_stdio(false); cin.tie(nullptr);

Author: Yes you need it
I tried without and it tle'd

As we are iterating greedily through the array, we can merge ($i$ and $j$) and ($i + n$ and $j + n$) if we don't want to swap. This works because if we swap neither or swap both the arrangement remains the same.
If we do want to swap we can merge ($i$ and $j + n$) and ($i + n$ and $j$). This works because if we swap exactly one of these, then the arrangement will change. 

After this we can construct the answer.

We can iterate over the entire grid. If it is a diaganal it will never change. If $a[i][j] == a[j][i]$ then we can manually set them to normal (since we didn't merge them in the disjoint set union when comparing them the states might have not been merged). If it is not a diaganol we can check whether $i$ and $j$ are merged in the disjoint set union. If they aren't then we will swap $a[i][j]$ and $a[j][i]$. Otherwise we will keep them the same.

## Implementation

**Time Complexity:** $\mathcal{O}(N^2)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

using vi = vector<int>;

> struct DSU {

Reviewer 2: also put this in a codesnip or smth

Author: done

	vi e;
	DSU(int n) {
		e = vi(n, -1);
	}
	
	int get(int x) {

> 		if (e[x] < 0) return x;

Reviewer 2: use braces _everywhere_

Author: fixed

---
id: cf-1713E
source: Codeforces Div 2
title: Cross Swapping
author: Aakash Gokhale
---

## Explanation

[Official Analysis](https://codeforces.com/blog/entry/105419)

Notice that $a[i][j]$ can only be swapped with $a[j][i]$ no matter which operations are performed.

For all $j$ such that $i < j$ (we only need to consider each pair once and keeping $i < j$ will ensure this), we can determine whether we want to swap them or keep them the same.

If $a[i][j] > a[j][i]$, we want to swap them. If $a[i][j] < a[j][i]$ we want to keep them the same. If they are equal it doesn't matter.

> However, it is possible that not all conditions can work together with each other and swapping something means we can't swap something else later on. Since we want the lexicographically smallest answer, the earlier indices matter more than later indices so we find optimal row swaps for earlier indices first.

Reviewer 2: > not all conditions can work together with each other

what does this mean exactly?

Author: swapping/keeping the same $a[i][j]$ with $a[j][i]$ means we can't swap/keep the same $a[k][l]$ with $a[l][k]$ because $i = k$, $i = l$, $j = k$, $j = l$, or some combination.

Does this work as a replacement?

Author: However, it is possible that not all conditions can work together with each other and swapping something/keeping something the same means we can't swap something else/keep something the same late on. For example it is impossible to swap $i$ and $j$, keep $j$ and $k$ the same, and keep $i$ and $k$ the same. We can first flip the $i$ith column to swap $i$ and $j$. We do nothing to $j$ or $k$ since we want to keep them the same. In order to keep $i$ and $k$ the same, we need to flip the $k$th column because we earlier swapped $i$ and two swaps will cancel out. However, swapping $k$ will mean that the second condition will fail and we have reached a scenario where we cannot satisfy all conditions.

Author: is this better?

Reviewer 2: at this point i'd just give an example

Author: For example it is impossible to swap $i$ and $j$, keep $j$ and $k$ the same, and keep $i$ and $k$ the same.

Isn't this already an example?

Reviewer 2: > swap $i$ and $j$

how can you swap $i$ and $j$? are these numbers? points?
instead of variables, just use concrete numbers, perhaps with the sample test case

Author: However, it is possible that not all conditions can work together with each other and swapping something/keeping something the same means we can't swap something else/keep something the same late on. For example it is impossible to swap a[i][j] and a[j][i], keep a[j][k] and a[k][j] the same, and keep a[i][k] and a[k][i] the same. We can first flip the ith column/row which satisfies the first condition. We do nothing to the jth or kth column/row since we want to keep them the same which satisfies the second condition. In order to keep a[i][k] and a[k][i] the same, we need to flip the kth column/row and two swaps (i and k) will cancel out. However, swapping the kth column/row will cause the second condition to now fail and we have reached a scenario where we cannot satisfy all conditions.

Author: I was unclear in my explanation. I was supposed to reference a[i][j], a[j][k], and a[i][k]

Reviewer 2: this is pretty hard to follow
i'd use the example input or make an example array & say "it's possible to swap 3 & 4, but then this would [something bad]"

Author: However, it is possible that not all conditions can work together with each other and swapping something/keeping something the same means we can't swap something else/keep something the same later on.
Consider to following input:
0 2 1
1 0 1
3 2 0
We want to first swap a[0][1] and a[1][0] because 1 is less than 2. We can do this by flipping the 0th row/col. This would give us the array:
0 1 3
2 0 1
1 2 0
Then we want to swap a[0][2] and a[2][0] because 1 is less than 3. To do this either we can either swap the 0th row/col or the 2nd row/col. However, swapping the 0th row/col would undo the efforts of our first operation so we will swap the 2nd row/col and get:
0 1 1
2 0 2
3 1 0
Now we want to swap a[1][2] and a[2][1] because 1 is less than 2 and we want the lexicographically minimum. However swapping the 1st row/col would mess up the swap we made in the first operation and swapping the 2nd row/col would mess up the swap we made in the second operation.
Thus, it is impossible to make all swaps optimally.

Is this better?

Reviewer 2: much much better

Author: fixed

		}
	}
	
	bool same(int a, int b) {
		return get(a) == get(b);
	}
	
	int size(int x) {
		return -e[get(x)];
	}
	
	bool unite(int x, int y) {
		x = get(x);
		y = get(y);
		if (x == y) {
			return false;
		}
		if (e[x] > e[y]) {
			swap(x, y);
		}
		e[x] += e[y];
		e[y] = x;
		return true;
	}
};
//EndCodeSnip

void solve() {
	int n;
	cin >> n;
	int a[n][n];
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			cin >> a[i][j];
		}
	}
	
	DSU dsu(2 * n);
	
	for (int i = 0; i < n; i++) {
		// Only check each pair once and diagnals don't matter
		for (int j = i + 1; j < n; j++) {
			// If they are the same we should not merge either way
			// because it could affect later merges
			if (a[i][j] == a[j][i]) {
				continue;
			} else if (a[i][j] > a[j][i]) {
				// Swap the elements if the states are not already merged the opposite way
				if (dsu.get(i) != dsu.get(j)) {
					dsu.unite(i, j + n);
					dsu.unite(j, i + n);
				}
			} else	{
				// Keep the elements the same if the states
				// are not already merged the opposite way
				if (dsu.get(i) != dsu.get(j + n)) {
					dsu.unite(i, j);
					dsu.unite(i + n, j + n);
				}
			}
		}
	}
	
	// Construct the answer

> 	int ans[n][n];

Reviewer 2: please use a vector

Author: vectors are too slow

Author: i got inconsistent results when i used a vector
sometimes tle and sometimes ac

---
id: cf-1713E
source: Codeforces Div 2
title: Cross Swapping
author: Aakash Gokhale
---

## Explanation

[Official Analysis](https://codeforces.com/blog/entry/105419)

Notice that $a[i][j]$ can only be swapped with $a[j][i]$ no matter which operations are performed.

> In addition, we want to lexicographically sort the array. Thus, we can iterate and greedily process the array.

Reviewer 2: > Thus, we can iterate and greedily process the array.

what does this mean

Author: Since we want the lexicographically smallest answer, the earlier indices matter more than later indices so we find optimal row swaps for earlier indices first.

Reviewer 2: then say that

Author: fixed

For all $j$ such that $i < j$ (we only need to consider each pair once and keeping $i < j$ will ensure this), we can determine whether we want to swap them or keep them the same.

If $a[i][j] > a[j][i]$, we want to swap them. If $a[i][j] < a[j][i]$ we want to keep them the same. If they are equal it doesn't matter.

> However, it is possible that not all conditions can work together with each other and swapping something means we can't swap something else later on.

Reviewer 2: so...how do we address this problem?

Author: fixed

	int size(int x) {
		return -e[get(x)];
	}
	
	bool unite(int x, int y) {
		x = get(x);
		y = get(y);
		if (x == y) {
			return false;
		}
		if (e[x] > e[y]) {
			swap(x, y);
		}
		e[x] += e[y];
		e[y] = x;
		return true;
	}
};
//EndCodeSnip

void solve() {
	int n;
	cin >> n;
	int a[n][n];
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			cin >> a[i][j];
		}
	}
	
	DSU dsu(2 * n);
	
	for (int i = 0; i < n; i++) {
		// Only check each pair once and diagnals don't matter
		for (int j = i + 1; j < n; j++) {
			// If they are the same we should not merge either way
			// because it could affect later merges
			if (a[i][j] == a[j][i]) {
				continue;
			} else if (a[i][j] > a[j][i]) {
				// Swap the elements if the states are not already merged the opposite way
				if (dsu.get(i) != dsu.get(j)) {
					dsu.unite(i, j + n);
					dsu.unite(j, i + n);
				}
			} else	{
				// Keep the elements the same if the states
				// are not already merged the opposite way
				if (dsu.get(i) != dsu.get(j + n)) {
					dsu.unite(i, j);
					dsu.unite(i + n, j + n);
				}
			}
		}
	}
	
	// Construct the answer
	int ans[n][n];
	for (int i = 0; i < n; i++) {
		for (int j = i; j < n; j++) {

> 			/* If a[i][j] is part of a diagonal or the pair is equal (a[i][j] == a[j][i]),
> 			 * manually set it to be the same as the original array
> 			 */

Reviewer 2: ```suggestion
			/*
			 * If a[i][j] is part of a diagonal or the pair is equal (a[i][j] == a[j][i]),
			 * manually set it to be the same as the original array
			 */
```

			if (i == j) {
				ans[i][j] = a[i][j];
			} else if (a[i][j] == a[j][i]) {
				ans[i][j] = ans[j][i] = a[i][j];
			} else { // Otherwise check whether the i and j states are merged
				if (dsu.get(i) == dsu.get(j)) {
					ans[i][j] = a[i][j];
					ans[j][i] = a[j][i];
				} else {
					ans[i][j] = a[j][i];
					ans[j][i] = a[i][j];
				}
			}
		}
	}
	

> 	// Print the answer

Reviewer 2: ```suggestion
```
comments need not state the obvious

			cout << ans[i][j] << " ";
		}

> 		cout << endl;

Reviewer 2: perhaps try using `\n` here instead of `endl`, see if that fixes the vector issue

Author: yea it fixed the issue

		}
		cout << endl;
	}
}

int main() { 
	ios::sync_with_stdio(false); cin.tie(nullptr);

> 	int T = 1; cin >> T;
> 	while (T--) {

Reviewer 2: ```suggestion
	int T;
	cin >> T;
	for (int t = 0; t < T; t++) {
```

For all $j$ such that $i < j$ (we only need to consider each pair once and keeping $i < j$ will ensure this), we can determine whether we want to swap them or keep them the same.

If $a[i][j] > a[j][i]$, we want to swap them. If $a[i][j] < a[j][i]$ we want to keep them the same. If they are equal it doesn't matter.

However, it is possible that not all conditions can work together with each other and swapping something means we can't swap something else later on.

We can create a disjoint set union to model this problem. The DSU will have $2n$ nodes where node $i$ represents the state where the operation is performed on the $i$th row and node $i + n$ represents the state where operation is not performed on the $i$th row.

As we are iterating greedily through the array, we can merge ($i$ and $j$) and ($i + n$ and $j + n$) if we don't want to swap. This works because if we swap neither or swap both the arrangement remains the same.
If we do want to swap we can merge ($i$ and $j + n$) and ($i + n$ and $j$). This works because if we swap exactly one of these, then the arrangement will change. 

> After this we can construct the answer.
> 
> We can iterate over the entire grid. If it is a diaganal it will never change. If $a[i][j] == a[j][i]$ then we can manually set them to normal (since we didn't merge them in the disjoint set union when comparing them the states might have not been merged). If it is not a diaganol we can check whether $i$ and $j$ are merged in the disjoint set union. If they aren't then we will swap $a[i][j]$ and $a[j][i]$. Otherwise we will keep them the same.

Reviewer 2: ```suggestion
After this we can construct the answer by iterating over the grid. If it is a diagonal, it will never change. If $a[i][j] == a[j][i]$ then we can manually set them to normal (since we didn't merge them in the disjoint set union when comparing them the states might have not been merged). If it is not a diagonal, we can check whether $i$ and $j$ are merged in the disjoint set union. If they aren't then we will swap $a[i][j]$ and $a[j][i]$. Otherwise, we will keep them the same.
```
what does "iterating over the entire grid"
if it's cell by cell, what does "being a diagonal" mean?

Author: iterating over grid =
for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
        // do stuff to a[i][j]
    }
}

diagonal = a[i][i]

Reviewer 2: say this then
"We construct the answer by iterating over each cell of the grid."
"If the cell's row index is equal to its column index (i.e. it's on the diagonal of the grid"

Author: fixed

---
id: cf-1713E
source: Codeforces Div 2
title: Cross Swapping
author: Aakash Gokhale
---

## Explanation

[Official Analysis](https://codeforces.com/blog/entry/105419)

Notice that $a[i][j]$ can only be swapped with $a[j][i]$ no matter which operations are performed.

For all $j$ such that $i < j$ (we only need to consider each pair once and keeping $i < j$ will ensure this), we can determine whether we want to swap them or keep them the same.

If $a[i][j] > a[j][i]$, we want to swap them. If $a[i][j] < a[j][i]$ we want to keep them the same. If they are equal it doesn't matter.

However, it is possible that not all conditions can work together with each other and swapping something/keeping something the same means we can't swap something else/keep something the same later on.
Consider to following input:

$0$ $2$ $1$
$1$ $0$ $1$

> $3$ $2$ $0$

Reviewer 2: did you check the formatting on the website? this seems kinda sus

Author: fixed

using vi = vector<int>;

> //BeginCodeSnip{}

Reviewer 2: ```suggestion
//BeginCodeSnip{DSU}
```
